packages/babel-helpers/src/helpers.js

```diff
@@ -427,46 +427,59 @@ helpers.inheritsLoose = defineHelper(`
 
 // Based on https://github.com/WebReflection/babel-plugin-transform-builtin-classes
 helpers.wrapNativeSuper = defineHelper(`
-  var _gPO = Object.getPrototypeOf || function _gPO(o) { return o.__proto__ };
-  var _sPO = Object.setPrototypeOf || function _sPO(o, p) { o.__proto__ = p; return o };
-  var _construct = (typeof Reflect === "object" && Reflect.construct) ||
-    function _construct(Parent, args, Class) {
-      var Constructor, a = [null];
-      a.push.apply(a, args);
-      Constructor = Parent.bind.apply(Parent, a);
-      return _sPO(new Constructor, Class.prototype);
-    };
-
-  var _cache = typeof Map === "function" && new Map();
+  function _gPO(o) {
+    _gPO = Object.getPrototypeOf || function _gPO(o) { return o.__proto__ };
+    return _gPO(o);
+  }
+  function _sPO(o, p) {
+    _sPO = Object.setPrototypeOf || function _sPO(o, p) { o.__proto__ = p; return o };
+    return _sPO(o, p);
+  }
+  function _construct(Parent, args, Class) {
+    _construct = (typeof Reflect === "object" && Reflect.construct) ||
+      function _construct(Parent, args, Class) {
+        var Constructor, a = [null];
+        a.push.apply(a, args);
+        Constructor = Parent.bind.apply(Parent, a);
+        return _sPO(new Constructor, Class.prototype);
+      };
+    return _construct(Parent, args, Class);
+  }
 
   export default function _wrapNativeSuper(Class) {
-    if (typeof Class !== "function") {
-      throw new TypeError("Super expression must either be null or a function");
-    }
+    var _cache = typeof Map === "function" && new Map();
```

Comment:
Wouldn't three-shakers remove this even if it was outside from the function, since `Map` is known to be pure?

Reply:
not every tool unfortunately have sets of pure builtins in their codebases (from what I know), I've tried to get rid of this statement with UglifyJS and couldnt make it work, cc @kzc

Reply:
Uglify regards `Map` as an unresolved global and won't reduce that expression. Even if it did recognize `Map` it would be conservative and assume that it can be overridden in the worst case:
```
$ echo '"use strict"; console.log(Map); Map=1; console.log(Map);' | node
[Function: Map]
1
```
Rollup, on the other hand, might be able to do it as it assumes an ES6 environment and that no builtins can be overridden.

Reply:
Then imho we should leave it in the form I've proposed.